File: lb1.2/python/zeidel.py
# -*- coding: utf-8 -*-
import numpy as np
from math import fabs
import os
import re

def check(A,size): # проверка условий
    numb = 0
    summ = 0
    for i in range(size):
        numb = fabs(A[i][i])
        for j in range(size):
            if i!=j:
                summ = summ + fabs(A[i][j])
        if numb < summ:
            print("Условие диагонального преобладания не выполнено!")
            raise SystemExit # выход!()
        summ = 0


def findX(method,A,b,X,tempX,size):
    if method == 1: # Для Якоби
        for i in range(size):
            X[i] = b[i]
            for j in range(size):
                if i != j:
                    X[i] = X[i] - A[i][j] * tempX[j]
            X[i] = X[i] / A[i][i]
    if method == 2: # для Гаусса-Зейделя
        for i in range(size):
            X[i] = b[i]
            for j in range(i):
                if i != j:
                    X[i] = X[i] - A[i][j] * X[j]
            for j in range(i,size):
                if i != j:
                    X[i] = X[i] - A[i][j] * tempX[j]
            X[i] = X[i] / A[i][i]

def calc(E,method,A,b,X,tempX,size,iterat): # Вычисления!
    Max = 0
    while 1:
        findX(method,A,b,X,tempX,size)
        iterat[0] += 1
        Max = fabs(X[0] - tempX[0])
        for i in range(1,size):
            if Max < fabs(X[i]-tempX[i]):
                Max = fabs(X[i]-tempX[i])
        if Max < E:
            return
        for i in range(size):
            tempX[i] = X[i]

iterat = [0]
size = int(raw_input("Введите размер матрицы:\n>>> "))
print("\n=========================================")
# E = float(raw_input("Введите точность:\n>>> "))
E = 0.001
A = np.zeros((size,size))
b = np.zeros(size)
X = np.zeros(size)
tempX = np.zeros(size)
filename = "matr.txt"
if os.path.exists(filename):#  Если файл существует
	file = open(filename)
	text = file.readlines()
	for i in range(size):
		arr = re.findall(r"[-]?\d+\.\d+|[-]?\d+", text[i])
		if len(arr) != size + 1:
			print("Размер матрицы меньше заданного числа!")
			raise SystemExit # выход!()
		else:
			for j in range(size):
				A[i][j] = arr[j]
			b[i] = arr[size]
else:
    print("Файл [matr.txt] не найден!\nПроверьте файлы программы...")
    raise SystemExit # выход!

check(A,size)
print("Матрица коеффициентов:")
print(A)
print("Матрица b:")
print(b)
calc(E,1,A,b,X,tempX,size,iterat)
print("\n=========================================")
print("Метод Якоби")
print("Решение:")
print(X)
print("Кол-во итераций: %d"%iterat[0])
for i in range(size):
    tempX[i] = 0
    X[i] = 0
iterat[0] = 0
calc(E,2,A,b,X,tempX,size,iterat)
print("\n=========================================")
print("Метод Гаусса-Зейделя")
print("Решение:")
print(X)
print("Кол-во итераций: %d"%iterat[0])

File: lb1.2/c++/zeidel.cpp
#include<iostream>
#include<cstdlib>
#include<cmath>
#include<fstream>
#include<ctime>
#include <iomanip>

#define filename "matr.txt"
// #define filename "mymatr.txt"

using namespace std;




// Функции для расчета =================================================


int check(double **A, int size){ // проверка условий
    int numb = 0;
    int summ = 0;
    int i, j;
    for (i = 0; i < size; i++) {
        numb = fabs(A[i][i]);
        for (j = 0; j < size; j++) {
            if (i != j)
                summ = summ + fabs(A[i][j]);
        }
        if (numb < summ){
            std::cout << "Условие диагонального преобладания НЕ выполнено!" << '\n';
            return 1;
        }
        summ = 0;
    }
    std::cout << "Условие диагонального преобладания выполнено!" << '\n';
    return 0;
}

int main(int argc, char const *argv[]) {
	int iterat = 0, size;
	int i, j, k;
    double E = 0.001; // точность
    double **A, *b, *X, **Z, *tempX;
    double tmp_doub1 = 0.0, tmp_doub2 = 0.0;
    float time0, time1;
    ifstream fin(filename);

	if (!fin.is_open()) { // проверка файла
		cerr << "Не удалось открыть файл [matr.txt]!\n";
		return 0;
    }
    cout << "Файл ["<< filename << "] успешно открыт!" << '\n';
    while (!fin.eof()) {
		fin >> tmp_doub1;
		size++;
	}
	size--;// из-за перехода на след строку(
	std::cout << "Всего эл-ов в матрице: "<< size << '\n';
	while (1) {
		tmp_doub1 = sqrt(size);
		i = modf(tmp_doub1, &tmp_doub2);
		// std::cout << "size = " << i << '\n';
		if (i == 0) {
			break;
		}
		size--;

	}
	size = static_cast<int>(tmp_doub2); // пеобразуем целое дабл в инт
	std::cout << "Размер матрицы: "<< size << '\n';
	fin.clear();// отчистим поток
	fin.seekg(0);// двинемся в начало файла

	std::cout << "\n=========================================" << '\n';

    A = new double* [size];// матр коэф
	for (i = 0; i < size; i++)
		A [i] = new double [size];

    Z = new double* [size];// матр коэф test
	for (i = 0; i < size; i++)
		Z[i] = new double [size];

    b = new double [size];// вектор
    X = new double [size];// корни
    tempX = new double [size];



    for (i = 0; i < size; i++) {
        for (j = 0; j < size; j++) {
            fin >> A[i][j];
        }
        fin >> b[i];
        tempX[i] = 0;
    }
	if (check(A, size) != 0)
        return 1;
	std::cout << "Матрица коеффициентов A и вектор b:" << '\n';
    for (i = 0; i < size; i++){
		for (j = 0; j < size; j++){
			cout << setw(5) << A[i][j];
		}
		cout << "  |" << setw(3) << b[i] << endl;
	}

    // ==================[ решение Якоби] ======================================================

    time0 = clock ();
    double Max = 0.0;
    while (1){
        for (i = 0; i < size; i++) {
            X[i] = b[i];
            for (j = 0; j < size; j++)
                if (i != j)
                    X[i] = X[i] - A[i][j] * tempX[j];
            X[i] = X[i] / A[i][i];
        }
        iterat += 1;
        Max = fabs(X[0] - tempX[0]);
        for (i = 1; i < size; i++) {
            if (Max < fabs(X[i]-tempX[i]))
                Max = fabs(X[i]-tempX[i]);
        }
        if (Max < E){
            break;
        }
        for (i = 0; i < size; i++)
            tempX[i] = X[i];
    }
    time1 = clock ();
	std::cout << "\n=========================================" << '\n';
	std::cout << "Метод Якоби" << '\n';
	std::cout << "Решение:" << '\n';
    for ( i = 0; i < size; i++) {
        std::cout << "  X"<< i+1 <<" = " << X[i] <<'\n';
        X[i] = 0;
        tempX[i] = 0;
    }
	std::cout << "Кол-во итераций: " << iterat << '\n';
    iterat = 0;
    cout << "На метод Якоби потрачено " << (time1 - time0) / 1000 << " секунд" << endl;
    std::cout << "\n=========================================" << '\n';
    // ==================[ решение Гаусса-Зейделя] ==================================================
    std::cout << "Метод Гаусса-Зейделя" << '\n';
	std::cout << "Решение:" << '\n';
    time0 = clock ();//
    Max = 0.0;
    while (1){
        for (i = 0; i < size; i++) {
            X[i] = b[i];
            for (j = 0; j < i; j++)
                if (i != j)
                    X[i] = X[i] - A[i][j] * X[j];
            for (j = i; j < size; j++)
                if (i != j)
                    X[i] = X[i] - A[i][j] * tempX[j];
            X[i] = X[i] / A[i][i];
        }

        iterat += 1;
        Max = fabs(X[0] - tempX[0]);
        for (i = 1; i < size; i++) {
            if (Max < fabs(X[i]-tempX[i]))
                Max = fabs(X[i]-tempX[i]);
        }
        if (Max < E){
            break;
        }
        for (i = 0; i < size; i++)
            tempX[i] = X[i];
    }
    time1 = clock ();//
    for ( i = 0; i < size; i++) {
        std::cout << "  X"<< i+1 <<" = " << X[i] <<'\n';
    }
    std::cout << "Кол-во итераций: " << iterat << '\n';
    cout << "На метод Гаусса-Зейделя потрачено " << (time1 - time0) / 1000 << " секунд" << endl;
    return 0;
}
